### PR TITLE
add inventory

### LINKED ISSUE #22: Suggestion: add inventory of already owned resources, that can be used
Example: I want to craft 20 bookshelves in minecraft. To do so I need books and planks. To craft books I need paper and leather, to craft paper I need sugar cane.

If I already own a couple of books and papers, I will need drastically less sugar cane.

Would be great to have that covered as well.

If you'd like the idea, I would try to implement it, as it would make a project idea, that I had for quite some time, pretty much obsolete.

### PR DESCRIPTION
**Pull request by mmodrow**
I added the functionality to keep track of already owned items by using an inventory of owned items.

This is stored in local storage and a textarea as JSON for local persistence and the option to import/export the inventory to other machines or browsers. The Textarea resides in a pane, that is hidden by default and can be shown via the "Inventory"-Button in the top menu.
Besides manipulating the localstorage or the textarea content as JSON, you can also use the double-click-menu on the item, where I added a number box to keep track of the selected item's inventory stock. The double-click-menu input updates the textarea, the in-memory-inventory and the localstorage inventory on change, so it's effectively 0-click interaction.

Items in the inventory will be used instead of "required" items wherever reasonable. E.g. if you have 3 planks and need 6 only 2 will be used, as they can only be crafted in sets of 4. Using all 3 would cause 1 [Extra] to pop up in the graph.
In the text instructions inventory usage will be denoted as "Take [amount] [item] from inventory." and in the base ingredients, owned and needed items will be listed separately, if any inventory items can be used.

If no inventory items are stored, nothing changes besides the Inventory-button in the top menu and the inventory amount headline+number input in the double click menu.

New UI:
image

UI with no new features in use:
image

This fixes #22 

I also added myself as author for minecraft for my contribution in https://github.com/AsherGlick/ResourceCalculator/issues/16 .
I also took the liberty, to fix some more typos, spelling mistakes and remove some random empty lines as well as make some optimizations, where I touched the code.

### CODE REVIEW

**Comment by AsherGlick** [comment]
Awesome! I will get a chance to review this over the weekend. I will say quickly though that I am a bit wary of introducing a raw JSON textbox as a UI with such a predominant button on the page. JSON looks very confusing to those who are unfamiliar with it. I think at the very least a clear and obvious "close" or "hide" button should be added to the JSON window in case someone accidentally clicks it without understanding what it is. Ideally the inventory button could be moved to somewhere less likely to be accidentally clicked too until it has a more user friendly UI (while still allowing experienced users to use it).

**Comment by mmodrow** [comment]
That sounds very reasonable.
The current version mainly aimed for minimal impact on the current UI style, while allowing intermediate and expert users an easy, robust and portable tool to copy&paste the inventory.

I'm very much open for suggestions.

I will have less time to carry this over the finish line, because of personal things demanding more time investment soon, but I gave you write access as well, in case you want to take a shot at the UI. I tried to use as much of your code and/or paradigms as possible, so that should not be too hard.

**Comment by mmodrow** [comment]
Does this look better?
image
Thinking about it... I think I could also get rid of the buttons below the textarea, by updating the in-memory-inventory on change and updating the textarea everytime I change the in-memory-inventory.

**Comment by mmodrow** [comment]
I did the auto-update-thing and removed the buttons.
image

If the user breaks something, they will be notified about it, so they can undo it.
image

File: core/calculator.html
Comment (mmodrow) [reply core/calculator.html:56]: image
Topmost: `╳` aka `&#9587;`
Middle: `⨉` aka `&#10761;`
Lowest: `X` aka `X`

**Comment by mmodrow** [comment]
I did my own code review and fixed all but one thins I came up with, that needs a 2nd opinion and I am curious to see, what you come up with.

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js:244]: Fixed

File: core/calculator.html
Comment (mmodrow) [reply core/calculator.html]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js]: Fixed

File: core/calculator.css
Comment (mmodrow) [reply core/calculator.css]: Fixed

**Comment by mmodrow** [comment]
@AsherGlick just to avoid misunderstandings:
Besides the "X"-question in my review, where I @ed you, this feels ready to me. I am awaiting your review.

File: core/calculator.js
Comment (AsherGlick) [inline core/calculator.js:456]: Is `used_from_inventory` necessary to pass in to `generate_chart()`?

File: core/calculator.js
Comment (AsherGlick) [inline core/calculator.js:857]: Will this ever be `!== undefined`? It looks like you populate `used_from_inventory` with the suffix-less name as queried a few lines up.

File: core/calculator.js
Comment (AsherGlick) [inline core/calculator.js:848]: I would have to follow through the code path again to check this myself but maybe you know as it is fresher in your mind. Will there ever be a case where `used_from_inventory[node_name.replace(inventory_label_suffix."")]` `!=` `node_quantities[node_name]`?

**Review by AsherGlick — CHANGES_REQUESTED** [review]
Overall it looks good, I like how you included the text box in the item dropdown menu, it is a good solution to the problem.

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js:456]: I use it in line 848 to get the output amount for the extra element for `[from Inventory]` items. I think it's more explicit if there is an extra column for those entries.

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js:848]: That is a very good question, that I will look into as soon, as I get back to my PC.

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js:857]: I'd have to look up the details, but from a glimpse this should be `undefined` by default and only be set for inventory items.

`used_from_inventory` should not be manipulated in this function at all.

File: core/calculator.html
Comment (mmodrow) [reply core/calculator.html]: &#9587; it is.

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js:857]: Just tried it in the debugger and yes, every item that is not in the inventory will hit this `if`.

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js:848]: It is always the case, if an item is needed and it will not be sourced from the inventory:
image

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js:315]: added the comment

File: core/calculator.html
Comment (mmodrow) [reply core/calculator.html:88]: NAR = no action required?

**Comment by mmodrow** [comment]
@AsherGlick  Finished responding to your comments.

File: core/calculator.js
Comment (AsherGlick) [reply core/calculator.js:849]: from your example it looks to me like they are equal in this case
`used_from_inventory["Chest"] = node_quantities["Chest [from Inventory]"] = 59`
`used_from_inventory["Hopper"] = node_quantities["Hopper [from Inventory]"] = 5`
`used_from_inventory["Redstone"] = node_quantities["Redstone [from Inventory]"] = 16`

so you should be able to remove that first if statement so
```
			if (node_name.endsWith(inventory_label_suffix)) {
				output = used_from_inventory[node_name.replace(inventory_label_suffix, "")];
			}
			else if (node_quantities[node_name] !== undefined) {
				output = node_quantities[node_name];
			}
			else {
				output = 0;
			}
```
becomes just
```
			if (node_quantities[node_name] !== undefined) {
				output = node_quantities[node_name];
			}
			else {
				output = 0;
			}
```

File: core/calculator.js
Comment (AsherGlick) [reply core/calculator.js:858]: Ah I see it is needed to make the height of the bar chart accurate. I feel like there is a cleaner way to do this but it is not worth holding up this pr for it and something that can maybe be looked at down the line.

File: core/calculator.js
Comment (AsherGlick) [reply core/calculator.js:457]: See comment below about that line, however it is still used in the block right after that so this will be necessary for now and I am not going to hold up the pr to see if there is a nicer way to handle this.

File: core/calculator.html
Comment (AsherGlick) [reply core/calculator.html:88]: Ah, work acronyms sneaking into my personal life... yes you guessed correctly.

**Comment by AsherGlick** [comment]
In pulling this pr to test my suggustion I found a couple of edge cases that cause undesired behavior.

**Bug 1**
Asking a raw resource to be pulled from the inventory causes the instruction text to fail to generate because it does not know what text to use for generating a "Raw Resource"
`Uncaught TypeError: recipe_type_functions[recipe_type] is not a function`

EG: `#stickypiston=1` with an inventory of `{"Cobblestone": 1}` causes this crash.
This can probably be solved by edge casing out `Raw Resource` from the instruction lookup.

**Bug 2**
If you have fewer resources then a single recipe requires for a single output the instructions will ask you to take negative items out of your inventory and the chart will not render correctly.

EG: `#stickypiston=1` with an inventory of `{"Oak Wood Planks": 1}`

I'm guessing the bug is here
```
let recipes_from_owned = Math.floor(owned / recipe_output);
let overshot_from_owned = owned % recipe_output;
let extra_from_produce = needed % recipe_output;

if (overshot_from_owned < extra_from_produce){
	recipes_from_owned--;
}
```
where `Math.floor(owned / recipe_output) == 0` and `overshot_from_owned < extra_from_produce`
```
let recipes_from_owned = Math.floor(1 / 4); // 0
let overshot_from_owned = 1 % 4; // 1
let extra_from_produce = 3 % 4; // 3

if (overshot_from_owned < extra_from_produce){ // 1<3
	recipes_from_owned--; // -1
}
```
Screenshot_2021-05-22 ResourceCalculator

**Comment by mmodrow** [comment]
Good job at finding these - I will look into them asap.

**Comment by mmodrow** [comment]
I pushed what should've fixed both.

File: core/calculator.js
Comment (mmodrow) [reply core/calculator.js:858]: Ah, I see what you mean. I will look into that, just to check for edge cases.

Comment (mmodrow) [reply core/calculator.js]: You were totally right about that one. Comitted & pushed your suggestion.

**Comment by mmodrow** [comment]
And I looked into your other suggestions as well, @AsherGlick.

**Review by AsherGlick — APPROVED** [review]

**Pull request merged**
